inst/templates/derivationchain/custom.js

```diff
@@ -368,8 +368,15 @@ function generateLegend(nodes) {
   var visibleProcessors = [];
   var addExternalLine = false;
   var legendContainer = $("#legendContainer");
-  $(legendContainer).html('<strong>Explanation</strong><br/><br/><span style="text-decoration:underline">Processor Types:</span> <br/>');
-
+  $(legendContainer).html('<strong>Explanation</strong><br/><br/><span style="text-decoration:underline">Dataset Processor Types:</span> <br/>');
+  
+  var flags = [], output = [], l = nodes.length, i;
+    for( i=0; i<l; i++) {
+      if( flags[nodes[i].data["location"]]) continue;
+      flags[nodes[i].data["location"]] = true;
+      output.push(nodes[i].data["location"]);
+    }
```

Comment:
I might be incorrectly understanding what this is doing but I think this process could mostly happen within the loop below this one?

If you create an array of added locations before the loop and add each node's location to that array if it's not already in the array and then after the loop you can have if(locations.length>1){addExternalLine = true;}

That way we won't be running through a loop of all of the nodes twice. The overall approach of your ticket seems good though!